src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,403 @@
+Worse, the UI might become completely unresponsive.
+
+[jank]: https://docs.flutter.dev/perf/rendering
+
+
+### Using a single-task background worker
+
+If your app’s UI becomes unresponsive due to a time-consuming computation —
+[parsing a large JSON file][json], for example —
+consider offloading that computation to a background worker.
+A common case, which is supported by the [Flutter `compute()` function][],
+is spawning an isolate that performs a computation and then exits:
+***[PENDING: describe the figure in a little more detail.]***
+
+[json]: https://docs.flutter.dev/cookbook/networking/background-parsing)
+ [Flutter `compute()` function]: https://docs.flutter.dev/cookbook/networking/background-parsing#4-move-this-work-to-a-separate-isolate
+
+![PENDING: alt text here](/guides/language/concurrency/images/isolate-bg-worker.png)
+
+A background worker can perform I/O
+(reading and writing files, for example), set timers, and more.
+It has its own memory (including its own copy of code) and
+doesn’t share any state with the main isolate.
+
+
+### Using the Isolate API
+
+Sometimes you might need to use the `Isolate` API directly.
+For example, you might use the `Isolate` API if
+your Dart Native app needs any of the following:
+
+* A background worker like what `compute()` supports,
+  but that might not be running in a Flutter app
+* A background worker that can
+  handle multiple requests or send multiple replies,
+  as shown in the following figure
+* Some other custom isolate behavior
+
+![PENDING: alt text here](/guides/language/concurrency/images/isolate-custom-bg-worker.png)
+<br>
+***[PENDING: The top arrow should be labeled "Create isolate: Isolate.spawn()"]***
+
+As the preceding figure shows,
+to use the `Isolate` API you start with the `Isolate.spawn()` method.
+Isolates then communicate by passing messages.
+
+{{site.alert.info}}
+  **API note:**
+  When an isolate calls `Isolate.spawn()`,
+  it sets up the new isolate with a copy of the calling isolate’s code.
+  In some special cases —
+  for example, when implementing tools like `dart pub run` —
+  you might need to use the much slower `Isolate.spawnUri()`,
+  which sets up the new isolate with a copy of the code at the specified URI.
+{{site.alert.end}}
+
+Each isolate message can deliver one object,
+which includes anything that’s transitively reachable from that object.
+Not all object types are sendable, and
+the send fails if any transitively reachable object is unsendable.
+For example, you can send an object of type `List<Object>` only if
+none of the objects in the list is unsendable.
+If one of the objects is, say, a `Socket`, then
+the send fails because sockets are unsendable.
+
+For information on the kinds of objects that you can send in messages,
+see the API reference documentation for the [`send()` method][].
+
+[`send()` method]: https://api.dart.dev/stable/dart-isolate/SendPort/send.html
+
+{{site.alert.tip}}
+  The `send()` method can be slow.
```

Comment:
I don't actually know the answer to this! Which is it?

Reply:
When used between isolates in different isolate groups(spawned using `Isolate.spawnUri`), both sender and receiver are doing similar O(number-of-objects-being-sent) work: serializing done by the sender and deserializing done by the receiver.

When used between isolates in the same isolate group(spawned using `Isolate.spawn`), the sender (normally) does more work. The sender validates the message being sent and copies mutable objects of the message for the receiver's consumption. Heap is shared between isolates, so the receiver just picks up a reference to the prepared copy that the sender give to the receiver. Only further processing that the receiver has to do is to canonicalize/recalculate hashes for certain objects. If no canonicalization is needed, receiver will do no additional work.
Sending immutable(or "shareable") data is done by passing a reference, so no copy is needed and is very fast.

Use of `Isolate.exit(port, object)` avoids creating a copy too because it is guaranteed that the sender exits and does not mutate sent data. Because of that `Isolate.exit` is faster for sender than regular `send`. Sender still has to validate that only allowed objects are being sent, so it still does O(number-of-objects) work.

`TransferableTypedData` is similar to regular `send`, in that its sender who pays higher price by creating Transferable wrapper around unstructured bytes. There is no object traversal happening though, it is just memory copy, so it is faster than sending structured objects. Receiver gets data in constant time too.

Reply:
Included in followup issue (#3720)